packages/svelte/src/compiler/migrate/index.js

```diff
@@ -586,6 +602,93 @@ const template = {
 		handle_events(node, state);
 		next();
 	},
+	SvelteComponent(node, { state, next, path }) {
+		let expression = state.str.original.substring(
+			/** @type {number} */ (node.expression.start),
+			node.expression.end
+		);
+
+		if (state.analysis.uses_props || state.analysis.uses_rest_props) {
+			let difference = 0;
+			// we need to migrate $$props and $$restProps by hand to update all the $$props and $$restProps in the expression
+			walk(
+				node.expression,
+				{},
+				{
+					Identifier(id_node, { path, next }) {
+						const parent = path[path.length - 1];
+						if (id_node.name !== '$$props' && id_node.name !== '$$restProps') return;
+
+						if (parent.type === 'MemberExpression' && parent.object !== id_node) return;
+
+						const props_name =
+							id_node.name === '$$props' || state.analysis.uses_props
+								? state.names.props
+								: state.names.rest;
+						const props_start =
+							/** @type {number} */ (id_node.start) - /** @type {number} */ (node.expression.start);
+
+						expression =
+							expression.substring(0, props_start + difference) +
+							props_name +
+							expression.substring(props_start + id_node.name.length + difference);
+
+						difference += props_name.length - id_node.name.length;
+						next();
+					}
+				}
+			);
+		}
+
+		if (
+			(node.expression.type !== 'Identifier' && node.expression.type !== 'MemberExpression') ||
+			!regex_valid_component_name.test(expression)
+		) {
+			let current_expression = expression;
+			expression = state.scope.generate('SvelteComponent');
+			let needs_derived = true;
+			for (let i = path.length - 1; i >= 0; i--) {
+				const part = path[i];
+				if (
+					part.type === 'EachBlock' ||
+					part.type === 'AwaitBlock' ||
+					part.type === 'IfBlock' ||
+					part.type === 'KeyBlock' ||
+					part.type === 'SnippetBlock' ||
+					part.type === 'Component'
```

Comment:
Also needs to take into account the "is a element with a named slot" case (e.g. `<div slot="foo" let:thisIsAComponent></div>`).

Reply:
Isn't that already covered by `Component` since the const will still be next to the `svelte:component` and an element can't have a slot if it's not inside a component?

But now that i write this i realize i need to add `SvelteComponent` itself probably in case of nested `SvelteComponent`'s

Reply:
You're right, if there's a named slot then there has to be a component parent - all good.